File: Data Structures & Algorithms/coin-change/submission-1.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, -2);
        return coins(coins, dp, amount);
    }

    public int coins(int[] coins, int[] dp, int current) {
        if (current == 0) return 0;
        if (dp[current] != -2) return dp[current];
        int min = Integer.MAX_VALUE;
        for (int coin : coins) {
            if (current - coin >= 0) {
                int sub = coins(coins, dp, current - coin);
                if (sub != -1) min = Math.min(min, sub + 1);
            }
        }
        dp[current] = min == Integer.MAX_VALUE ? -1 : min;
        return dp[current];
    }
}

File: Data Structures & Algorithms/coin-change-ii/submission-1.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        vector<vector<int>> dp(coins.size(), vector<int>(amount, -1));
        return combination(dp, coins, 0, 0, amount);
    }

    int combination(vector<vector<int>>& dp, vector<int>& coins, int start, int current, int target) {
        if (current == target) return 1;
        if (current > target) return 0;
        if (dp[start][current] != -1) return dp[start][current];
        int possible = 0;
        for (int i = start; i < coins.size(); i++) {
            int coin = coins[i];
            if (current + coin > target) continue;
            possible += combination(dp, coins, i, current + coin, target);
        }
        dp[start][current] = possible;
        return dp[start][current];
    }
};
